### PR TITLE
fix(UPM-15878): adding first version of GCP setup-csp

### PR DESCRIPTION
**Pull request by valeriodelsarto**

### CODE REVIEW

File: gcp/biganimal-csp-setup [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+      ;;
+  esac
+done
+
+download_dependencies()
+{
+  # a shell func to download all the dependecies of this script before starting to execute
+  # Download the latest biganimal role template or overwrite it
+  template_file=biganimal_GCP_baserole.yaml
+  curl https://raw.githubusercontent.com/EnterpriseDB/cloud-utilities/upm-15878/gcp/${template_file} -o "${CURRENT_PATH}"/${template_file}
```
Comment (valeriodelsarto) [inline gcp/biganimal-csp-setup:166]: branch needs to be set to `main` after https://github.com/EnterpriseDB/cloud-utilities/pull/115 is merged

File: gcp/biganimal-csp-setup [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+    --filter "name:biganimal_${suffix_role_name}" --format json | jq '. | length')
+  if [ "$role_exists" -gt 0 ]; then
+    # a custom role with the same name already exists
+    echo -e "Updating GCP custom role ${GREEN}biganimal_${suffix_role_name}${NC} within GCP project ${GREEN}${project}${NC}..."
+
+    # updating the GCP custom role
+    custom_role_id=$(gcloud iam roles list --project "$project" \
+      --filter "name:biganimal_${suffix_role_name}" --format json | jq -r '.[].name')
+
+    gcloud iam roles update "${custom_role_id##*/}" --project "$project" \
```
Comment (fisher046) [inline gcp/biganimal-csp-setup:205]: With this command, we can update existing role with new permissions. For update-csp, we can reuse this script, am I right?

File: gcp/biganimal-csp-setup [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+  else
+    # the service account already exists, going to use it
+    echo -e "GCP service account ${GREEN}biganimal-${suffix_sa_name}${NC} already exists within GCP project ${GREEN}${project}${NC}..."
+  fi
+  service_account_id=$(gcloud iam service-accounts list --project "$project" \
+    --filter "name:biganimal-${suffix_sa_name}" --format json | jq -r '.[].email')
+
+  # assigning the custom BigAnimal Role to the service account for the given GCP project
+  echo -e "Allowing GCP service account ${GREEN}biganimal-${suffix_sa_name}${NC} to use the GCP project ${GREEN}${project}${NC}..."
+  gcloud projects add-iam-policy-binding "$project" \
```
Comment (fisher046) [inline gcp/biganimal-csp-setup:239]: What about the output if the role is already bound to the service account?

File: gcp/biganimal-csp-setup (hunk 2 above)
Comment (valeriodelsarto) [reply gcp/biganimal-csp-setup:205]: yes you're right, this works. So we can re-use the same script to update the existing custom role with the new content, and I've already tested it on the dev project

File: gcp/biganimal-csp-setup (hunk 3 above)
Comment (valeriodelsarto) [reply gcp/biganimal-csp-setup:239]: it works even if this `iam-policy-binding` already exists, so if it is not already existing it gets added, and if it already exists it gets updated. I've already tested it on the dev project

**Comment by jasonparton** [comment]
A few questions...

1. For documentation purposes, what is the actual role required to perform run the csp-setup script (which will be called on by the new BigAnimal CLI (v2.0) `**cloud-provider setup**` command?
Is it _roles/iam.serviceAccountCreator_, _roles/iam.serviceAccountKeyAdmin_, _roles/iam.roleAdmin_, or _roles/resourcemanager.projectIamAdmin_?  It may be rare that we have someone with the permissions of _roles/owner_ be available for onboardings.

2. Safe to assume the CLI team will be able to have the same BA CLI command to run in interactive mode?
3. Where do we stand with the preflight script?

**Comment by jasonparton** [comment]
> This is documented [here](https://github.com/EnterpriseDB/cloud-utilities/pull/116/files#diff-cfa362f7c8299c9ca2182841854f072717fc48089d6fe1e84108ecb5a33eddf0R62-R68). Let me know if you prefer to have them stated also somewhere else!
Nope.  Just seeking confirmation that those were indeed the GCP role(s) that could run the setup, and that a combination of the roles were required.  As for having it documented elsewhere, that will be necessary with the Docs team begins the drafts of the GCP content.  But I know that work has yet to begin.

**Pull request merged**